File: Leetcode/11. Container With Most Water/11_Container_Water.java
class Solution {
    public int maxArea(int[] height) {
        int n = height.length;
        if (n == 0)
            return 0;
        int max_area = 0;
        int left = 0, right = n - 1;
        while (left < right) {
            max_area = Math.max(max_area, Math.min(height[left], height[right]) * (right - left));
            if (height[left] < height[right])
                left++;
            else
                right--;
        }
        return max_area;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] height = { 1, 8, 6, 2, 5, 4, 8, 3, 7 };
        System.out.println(sol.maxArea(height));
    }
}

File: Leetcode/11. Container With Most Water/11_Container_Water.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution
{
public:
    int maxArea(vector<int> &height)
    {
        int max_area = 0;
        int left = 0, right = height.size() - 1;
        while (left < right)
        {
            max_area = max(max_area, min(height[left], height[right]) * (right - left));
            if (height[left] < height[right])
                left++;
            else
                right--;
        }
        return max_area;
    }
};

int main()
{
    Solution solution;
    vector<int> height = {1, 8, 6, 2, 5, 4, 8, 3, 7};
    cout << solution.maxArea(height) << endl;
    return 0;
}